dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+                                         self.mdelta, self.nbins, self.padding,
+                                         self.joint_grad)
+        elif dim ==3:
+            _joint_pdf_gradient_dense_3d(theta, transform, static, moving,
+                                         grid_to_space, mgradient, smask, mmask,
+                                         self.smin, self.sdelta, self.mmin,
+                                         self.mdelta, self.nbins, self.padding,
+                                         self.joint_grad)
+        else:
+            msg = 'Only dimensions 2 and 3 are supported. '+str(dim)+' received'
+            raise ValueError(msg)
+
+
+    def update_gradient_sparse(self, theta, transform, sval, mval,
+                               sample_points, mgradient):
+        r''' Computes the Gradient of the joint PDF w.r.t. transform parameters
+
+        Computes the vector of partial derivatives of the joint histogram w.r.t.
+        each transformation parameter.
+
+        Parameters
+        ----------
+        theta: array, shape (n,)
+            parameters to compute the gradient at
+        transform: instance of Transform
+            the transformation with respect to whose parameters the gradient
+            must be computed
+        sval: array, shape (m,)
+            sampled intensities from the static image at sampled_points
+        mval: array, shape (m,)
+            sampled intensities from the moving image at sampled_points
+        sample_points: array, shape (m, 3)
+            coordinates (in physical space) of the points the images were
+            sampled at
+        mgradient: array, shape (m, 3)
+            the gradient of the moving image at the sample points
+        '''
+        dim = sample_points.shape[1]
+        n = theta.shape[0]
+        nbins = self.nbins
+
+        if (self.joint_grad is None) or (self.joint_grad.shape[2] != n):
+            self.joint_grad = np.ndarray(shape = (nbins, nbins, n))
+
+        if dim == 2:
+            _joint_pdf_gradient_sparse_2d(theta, transform, sval, mval,
+                                          sample_points, mgradient, self.smin,
+                                          self.sdelta, self.mmin, self.mdelta,
+                                          self.nbins, self.padding,
+                                          self.joint_grad)
+        elif dim ==3:
+            _joint_pdf_gradient_sparse_3d(theta, transform, sval, mval,
+                                          sample_points, mgradient, self.smin,
+                                          self.sdelta, self.mmin, self.mdelta,
+                                          self.nbins, self.padding,
+                                          self.joint_grad)
+        else:
+            msg = 'Only dimensions 2 and 3 are supported. '+str(dim)+' received'
+            raise ValueError(msg)
+
+
+    def update_mi_metric(self, update_gradient=True):
```

Comment:
It always computes the metric value, but the gradient will only be computed if "update_gradient" is True. This may save some time if the optimization algorithm doesn't need to compute the gradient but only its value at some point of the optimization (for example during line-search to determine the step size along the previously computed descent direction)